### PR TITLE
Propagate package to ContractInstance

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
this advances the state of #17995

### CODE REVIEW

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala [hunk 1]
```diff
@@ -29,7 +35,7 @@ import java.util
 import scala.language.implicitConversions
 import scala.util.{Failure, Try}
 
-class SBuiltinTestV1 extends SBuiltinTest(LanguageMajorVersion.V1)
+//class SBuiltinTestV1 extends SBuiltinTest(LanguageMajorVersion.V1)
```
Comment (paulbrauner-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala:38]: Is this left commented out on purpose?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala [hunk 2]
```diff
@@ -194,15 +197,21 @@ object Value {
...
-    def apply(template: Identifier, arg: VersionedValue): VersionedContractInstance =
-      arg.map(ContractInstance(template, _))
+    def apply(
+        packageName: Option[Ref.PackageName] = None,
+        template: Identifier,
+        arg: VersionedValue,
+    ): VersionedContractInstance =
+      arg.map(ContractInstance(packageName, template, _))
 
+    @deprecated("use the version with 3 argument", since = "2.9.0")
```
Comment (paulbrauner-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala:207]: Is it still used anywhere? I see that you've replaced a bunch of occurrences in this PR.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala [hunk 3]
```diff
@@ -194,15 +197,21 @@ object Value {
   type VersionedContractInstance = transaction.Versioned[ContractInstance]
 
   object VersionedContractInstance {
-    def apply(template: Identifier, arg: VersionedValue): VersionedContractInstance =
-      arg.map(ContractInstance(template, _))
+    def apply(
+        packageName: Option[Ref.PackageName] = None,
```
Comment (paulbrauner-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala:201]: Is the plan to eventually remove the default value? If not then maybe we should also add TODOs to the use sites in Canton or we may forget to update them. Or will there be cases where it makes sense not to provide a package name?

**Review by paulbrauner-da — APPROVED** [review]

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala:38]: good catch.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala:207]: Maybe in canton.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala:201]: add a todo.

**Pull request merged**